pac4j-core/src/main/java/org/pac4j/core/profile/Pac4JPrincipal.java

```diff
@@ -0,0 +1,61 @@
+package org.pac4j.core.profile;
+
+import java.security.Principal;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+import org.pac4j.core.util.CommonHelper;
+
+public class Pac4JPrincipal implements Principal {
+
+    private final CommonProfile profile;
+
+    private final Set<String> roles;
+
+    public Pac4JPrincipal(LinkedHashMap<String, CommonProfile> profiles) {
+        Optional<CommonProfile> optProfile = ProfileHelper.flatIntoOneProfile(profiles);
+        if (!optProfile.isPresent()) {
+            throw new IllegalArgumentException();
+        }
+        this.profile = optProfile.get();
+        Set<String> rs = new HashSet<>();
+        profiles.values().stream().forEach(p -> rs.addAll(p.getRoles()));
```

Comment:
We gather the roles from all profiles, but we all return the identifier of only one profile.

Reply:
Should we return instead only the roles of the one profile selected by `flatIntoOneProfile`?

Reply:
Finally, I think it's better to return all roles. Keep things that way.

Reply:
That raises the issue of why we don't do that when we return a profile using `ProfileHelper.flatIntoOneProfile()` in all pac4j... Maybe it depends where it is used.

Reply:
For me, it's not the same use case: using `flatIntoOneProfile`, we choose one profile out of the several ones we may have. It's a complete profile with roles and identifier. Here, we have no choice, we try to get as much as we can from the pac4j information: all roles, but only one identifier as we can only return one name.

I think we should not keep a profile variable in this principal but only a name property, built in the constructor from the profile identifier.
We could even imagine to have this name built from all profile identifiers.

Reply:
Yep, that makes sense for not keeping the profile.
I don't know about the name being a compound of the all the profile's name thought (because usually I think people would be expecting the name to be a username for example).

I think we should go back to the reason why we are building a principal. Right now it's not clear when this information is extracted and why it is needed, so it's not clear what should be kept or not. I will try to think a bit about it in the context of jax-rs-pac4j and get back on that :)

Reply:
A compound of all the profile's names was just an example, it clearly makes more sense to have a username.

I would keep all roles, the id of the `flatIntoOneProfile`-returned profile and keep the id property instead of the whole profile.

Reply:
Actually, I just realised that the `getRoles()` method is something I introduced but is not part of the `Principal` interface (I stored it there because it made sense in jax-rs-pac4j).

So in the end, the principal is just a name (and an `implies` relation with a `Subject`)… and thus, I wonder if we shouldn't do thing the other way around: a pac4j profile IS-A Principal and should simply implements the interface.

What do you think?

Reply:
Actually, it makes a lot more sense to have it that way.

Reply:
Hm, actually, that's not so obvious, because implementing Principal has strong implications in terms of semantic:
 - first there is the method names that are very generic and will clash with CommonProfile
 - also there is the hashCode/equals method that may not be exactly what we would expect for a UserProfile

So in the end, I think I will continue with the idea of having a separate class… I'm going to rework this branch and you will tell me what you think :)

Reply:
ok, it's done